docs/HTTP_V2_SPEC.md

```diff
@@ -1,106 +1,236 @@
+must serialize the requests: it must wait for a response to the previous request
+before sending next request.
 
-## Version check
+The server can also optionally specify a different URL that the client should
+use for the requests on the stream. This can be used to ensure that stream
+requests are "sticky" and reach the same server.
 
-```
+The server will close streams after a short period of inactivity, to make sure
+that abandoned streams don't accumulate on the server.
+
+## Check support for version 2
+
+```typescript
 GET /v2
 ```
 
-A server that supports HTTP API v2 must return a success response for requests
-on path `/v2`. This can be used as a crude and unreliable mechanism for clients
-to fall back to version 1 of the protocol if the server does not support version
-2.
+If the server supports this version of the HTTP API, it should return a 2xx
+response for a GET request on `/v2`. This can be used as a crude version
+negotiation mechanism by the client.
 
-## Execute a statement
+## Execute requests on a stream
 
-```
-POST /v2/execute
+```typescript
+POST /v2/stream
 
 -> {
-    "stmt": Stmt,
+    "baton": string | null,
+    "requests": Array<StreamRequest>,
 }
 
 <- {
-    "result": StmtResult,
+    "baton": string | null,
+    "base_url": string | null,
+    "results": Array<StreamResult>
+}
+
+type StreamResult =
+    | StreamResultOk
+    | StreamResultError
+
+type StreamResultOk = {
+    "type": "ok",
+    "response": StreamResponse,
+}
+
+type StreamResultError = {
+    "type": "error",
+    "error": Error,
 }
 ```
 
-The `execute` endpoint receives a statement and returns the result of executing
-the statement. The `Stmt` and `StmtResult` structures are from the Hrana
-protocol. The semantics of this endpoint is the same as the `execute` request in
-Hrana.
+The `stream` endpoint is used to execute requests on a stream. `baton` in the
+request specifies the stream. If the client sets `baton` to `null`, the server
+should create a new stream.
+
+Server responds with another `baton` value in the response. If the `baton` value
+in the response is `null`, it means that the server has closed the stream. The
+client must use this value to refer to this stream in the next request (the
+`baton` in the response should be different from the `baton` in the request).
+This forces the client to issue the requests serially: it must wait for the
+response from a previous `stream` request before issuing another request on the
+same stream.
+
+The server should ensure that the `baton` values are unpredictable and
+unforgeable, for example by cryptographically signing them.
+
+If the `base_url` in the response is not `null`, the client should use this URL
+when sending further requests on this stream. If it is `null`, the client should
+use the same URL that it has used for the previous request. The `base_url`
+must be an absolute URL with "http" or "https" scheme.
```

Comment:
as per above, you can force the url and avoid base, no composition needs to be done, just get the opaque value and use it as the url. No headers or url assembling is needed.

you can use it in shell scripts and barebone clients with ease

Reply:
The client needs to do URL resolution anyway: the user passes an absolute URL like "https://my.turso.io" and the client must add "v2/stream" to obtain "https://my.turso.io/v2/stream". When the server returns a `base_url`, the client just replaces the user-provided URL with the `base_url`.

Reply:
(and there are no headers involved, everything is passed in request and response bodies as JSON)

Reply:
and what's the advantage of base_url over a fully ready control url, with it all. And it's not optional, so it's consistent, can (should be) the same.

Reply:
The advantage is that the server does not have to know its URL.

Reply:
But that's the easiest part. While doing that, it also helps the reverse proxy to properly connect.